README.md
[![Discord Support Server](https://img.shields.io/discord/690172711770652779?color=5865F2&label=support%20server&logo=discord&logoColor=fff&style=for-the-badge)](https://discord.gg/MmmEYefD8k)

# VoiceNotify

VoiceNotify is a Discord bot that notifies people when a voice chat is taking place. [Add to Discord](https://discord.com/oauth2/authorize?client_id=712670038267789352&scope=bot&permissions=150528)

Self-hosting : [Stable](https://github.com/vpctorr/DiscordBots/tree/main/packages/VoiceNotify) • [Dev](https://github.com/vpctorr/DiscordBots/tree/dev/packages/VoiceNotify)

### How to use the bot ?

Bot commands to enable & disable voice chat notifications (administrators only) :

`@VoiceNotify enable [threshold] [roles]`
Enables voice chat notifications for the voice channel you are in, alerts will be sent to the channel where this command is executed.

Optional : [threshold] to trigger an alert defaults to 5 people ; [roles] will be mentioned when the alert is sent.

`@VoiceNotify disable`
Disables voice chat notifications for the voice channel you are in.

---

# MakePDF

MakePDF is a Discord bot that automatically converts word processing, spreadsheet and presentation documents to PDF. [Add to Discord](https://discord.com/oauth2/authorize?client_id=932278614911766599&scope=bot&permissions=52224)

Self-hosting : [Stable](https://github.com/vpctorr/DiscordBots/tree/main/packages/MakePDF) • [Dev](https://github.com/vpctorr/DiscordBots/tree/dev/packages/MakePDF)

### Common formats are supported :

- DOC ; DOCX ; ODT ; RTF
- XLS ; XLSX ; OSD ; CSV
- PPT ; PPS ; PPTX ; PPSX ; ODP

### How does this work ?

Simply add MakePDF to your server. It will automatically convert all supported files to PDF, in all the channels it has access to.

We use the LibreOffice API to convert your files. Files are only temporarily transitting via our server but are not kept after the task is done.


CHANGELOG.md
# Changelog

## [1.7.1](https://github.com/vpctorr/DiscordBots/compare/discordbots-v1.7.0...discordbots-v1.7.1) (2023-01-24)


### 📦 Chores

* fix eslint ([cdf0200](https://github.com/vpctorr/DiscordBots/commit/cdf0200264d05428051e1673a40e06fcb410ed5c))
* update dependencies ([cc07378](https://github.com/vpctorr/DiscordBots/commit/cc07378303480b4808e58f0bb8da3d63f1c16ff4))


### 🔨 Improvements

* add a try/catch block around stats for safety ([3be4fc4](https://github.com/vpctorr/DiscordBots/commit/3be4fc473ecede8c77be0b0e099f4d5ae8ba88a5))
* show environment info in debug ([d56be2f](https://github.com/vpctorr/DiscordBots/commit/d56be2fa98a49f170d547efa3e323e87bf96d92e))
* **vn:** fixed crashes & bot not working ([3b7ea56](https://github.com/vpctorr/DiscordBots/commit/3b7ea562742135801338371a341a9623a41406dc))


### 🚚 Integration

* update create-release.yml ([bdc1919](https://github.com/vpctorr/DiscordBots/commit/bdc19196399d6a5a6bbb7589e5b9f3167a281577))

## [1.7.0](https://github.com/vpctorr/DiscordBots/compare/discordbots-v1.6.0...discordbots-v1.7.0) (2022-05-23)


### 🚚 Integration

* differentiate package changelogs ([c033457](https://github.com/vpctorr/DiscordBots/commit/c033457664ea530c0ddbfc63ba6a5169800892b8))


### 📚 Documentation

* updated links & added debug info ([29b832f](https://github.com/vpctorr/DiscordBots/commit/29b832fae6d7fd8ecffe1ef950fbfc5926518ae0))


### ✨ Features

* convert VoiceNotify to discord.js v13 ([#49](https://github.com/vpctorr/DiscordBots/issues/49)) ([c4b5ee3](https://github.com/vpctorr/DiscordBots/commit/c4b5ee336de40a4b4d4fb600216c3d2fdb835d18))


### 🧹 Refactor

* move bots logic to src ([c46290f](https://github.com/vpctorr/DiscordBots/commit/c46290f691d668ae30f23b2332db0a83accd5edb))

## [1.6.0](https://github.com/vpctorr/DiscordBots/compare/discordbots-v1.5.2...discordbots-v1.6.0) (2022-02-20)


### 📚 Documentation

* fix readme env var name ([eddbacb](https://github.com/vpctorr/DiscordBots/commit/eddbacb93cc4af0da1ee899af8921118c0e6648e))


### 📦 Chores

* update dependencies ([b01d669](https://github.com/vpctorr/DiscordBots/commit/b01d6695971908a5c7f2042877d3f1d0cf744152))
* updated dependencies ([fecf6fe](https://github.com/vpctorr/DiscordBots/commit/fecf6fedc92f5cbad7aaa306585266c60b0f3e7e))


### 👷 Runtime

* add postversion standalone script ([18fd8c2](https://github.com/vpctorr/DiscordBots/commit/18fd8c2d801e54d9d8abcae586495ad2c1da2c06))
* convert to npm workspaces ([2dfcac7](https://github.com/vpctorr/DiscordBots/commit/2dfcac7ebd50ce50727fea3df11977b89ae8a6ef))
* finish migrating to npm workspaces ([40f0d14](https://github.com/vpctorr/DiscordBots/commit/40f0d14a8b2c1c74c2ab2541616491c937b18416))
* remove postversion (will use github action) ([bedaa16](https://github.com/vpctorr/DiscordBots/commit/bedaa16ca158f9be273b3e59e24e7445cd196650))
* updated json modules implementation ([fd70b05](https://github.com/vpctorr/DiscordBots/commit/fd70b05d8e5ab7be1e5aec222214b5d86056ea71))


### ✨ Features

* **vn:** clear timeout with debug command ([85bf11c](https://github.com/vpctorr/DiscordBots/commit/85bf11c91f57c9a1224392dfa01bb78d6b3adbdd))


### 🔨 Improvements

* **mk:** soffice convert use path env var ([#20](https://github.com/vpctorr/DiscordBots/issues/20)) ([2ba7f80](https://github.com/vpctorr/DiscordBots/commit/2ba7f8026f641ab3abb9cf97650379586f7ebd6c))
* **mk:** use unicode emojis ([da08ee8](https://github.com/vpctorr/DiscordBots/commit/da08ee841872dcabc5e16581dd7d653578bf9370))
* use heroku slug as version for staging deploy ([3356b76](https://github.com/vpctorr/DiscordBots/commit/3356b762d1fdd21ffc9897fd61d48168527a39e6))
* **vn:** use unicode emoji for microphone ([#36](https://github.com/vpctorr/DiscordBots/issues/36)) ([d1d9d82](https://github.com/vpctorr/DiscordBots/commit/d1d9d821dcc1251b294fd47b4aa74272d2b2f4f8))


### 🚚 Integration

* new versioning tool ([23eb7c0](https://github.com/vpctorr/DiscordBots/commit/23eb7c02d2ea33c319029ba72d2b5a4a89f55c66))


index.js
require('dotenv').config()
const Discord = require('discord.js');
const https = require('https');
const libre = require('libreoffice-convert');

// Create an instance of a Discord client
const client = new Discord.Client();

// Bot will be able to send & receive to Discord only after this
client.on('ready', () => {
    console.log('Bot started & ready to answer !');
    client.user.setStatus('Converting files for you ⚡');
});

// On message received
client.on('message', msg => {

    try {

        if (msg.author.id == client.user.id) return

        const Attachment = (msg.attachments).array();

        if (Attachment === undefined || Attachment === null) return;

        Attachment.forEach((attachment) => {

            if (attachment === undefined || attachment === null) return;

            const fileName = attachment.name;

            const formats = process.env.FORMATS.split(',');
            const extension = fileName.substring(fileName.lastIndexOf(".") + 1);

            if (formats.includes(extension)) {

                https.get(attachment.url, res => {

                    const bufs = [];

                    res.on('data', function (chunk) {
                        bufs.push(chunk)
                    });

                    res.on('error', function (err) {
                        console.log(`Error during HTTP request : ${err.message}`);
                    });

                    res.on('end', function () {

                        const fileData = Buffer.concat(bufs);

                        libre.convert(fileData, ".pdf", undefined, (err, pdfData) => {

                            if (err) {
                                msg.channel.send(`Sorry, the conversion has failed :(`);
                                console.log(`Error converting file : ${err}`);
                            }

                            const newName = fileName.substring(0, fileName.lastIndexOf(".")) + ".pdf";
                            const newAttachment = new Discord.MessageAttachment(pdfData, newName);
                            msg.channel.send(`Here is your converted PDF file !`, newAttachment);

                        });

                    });

                });

            }

        });

    } catch (ex) {

        console.log(`Error : ${ex}`);

    }

});

client.login(process.env.TOKEN);

MakePDF/convert.js
'use strict'

const async = require('async')
const { execFile } = require('child_process')
const fs = require('fs')
const path = require('path')
const tmp = require('tmp')

const convertWithOptions = (document, format, filter, options, callback) => {
  const tmpOptions = (options || {}).tmpOptions || {}
  const asyncOptions = (options || {}).asyncOptions || {}
  const tempDir = tmp.dirSync({
    prefix: 'libreofficeConvert_',
    unsafeCleanup: true,
    ...tmpOptions
  })
  const installDir = tmp.dirSync({
    prefix: 'soffice',
    unsafeCleanup: true,
    ...tmpOptions
  })
  return async.auto(
    {
      soffice: (callback) => {
        let paths = []
        switch (process.platform) {
          case 'darwin':
            paths = ['/Applications/LibreOffice.app/Contents/MacOS/soffice']
            break
          case 'linux':
            paths = [
              '/app/vendor/libreoffice/usr/bin/libreoffice7.2',
              '/usr/bin/libreoffice',
              '/usr/bin/soffice',
              '/snap/bin/libreoffice'
            ] //
            break
          case 'win32':
            paths = [
              path.join(process.env['PROGRAMFILES(X86)'], 'LIBREO~1/program/soffice.exe'),
              path.join(process.env['PROGRAMFILES(X86)'], 'LibreOffice/program/soffice.exe'),
              path.join(process.env.PROGRAMFILES, 'LibreOffice/program/soffice.exe')
            ]
            break
          default:
            return callback(new Error(`Operating system not yet supported: ${process.platform}`))
        }

        return async.filter(
          paths,
          (filePath, callback) => fs.access(filePath, (err) => callback(null, !err)),
          (err, res) => {
            if (err || res.length === 0) {
              return callback(new Error('Could not find soffice binary'))
            }

            return callback(null, res[0])
          }
        )
      },
      saveSource: (callback) => fs.writeFile(path.join(tempDir.name, 'source'), document, callback),
      convert: [
        'soffice',
        'saveSource',
        (results, callback) => {
          let command = `-env:UserInstallation=file://${installDir.name} --headless --convert-to ${format}`
          if (filter !== undefined) {
            command += `:"${filter}"`
          }
          command += ` --outdir ${tempDir.name} ${path.join(tempDir.name, 'source')}`
          const args = command.split(' ')
          return execFile(results.soffice, args, callback)
        }
      ],
      loadDestination: [
        'convert',
        (results, callback) =>
          async.retry(
            {
              times: asyncOptions.times || 3,
              interval: asyncOptions.interval || 200
            },
            (callback) => fs.readFile(path.join(tempDir.name, `source.${format.split(':')[0]}`), callback),
            callback
          )
      ]
    },
    (err, res) => {
      tempDir.removeCallback()
      installDir.removeCallback()

      if (err) {
        return callback(err)
      }

      return callback(null, res.loadDestination)
    }
  )
}

const convert = (document, format, filter, callback) => {
  return convertWithOptions(document, format, filter, {}, callback)
}

module.exports = {
  convert,
  convertWithOptions
}


MakePDF/makepdf.js
/* eslint-disable eqeqeq */

import { Client, WebhookClient, MessageAttachment, MessageEmbed } from 'discord.js'
import { get, request } from 'https'

import { convert } from './convert.js'
import info from './package.json'
const client = new Client()
const hook = new WebhookClient(process.env.MAKEPDF_WEBHOOK_ID, process.env.MAKEPDF_WEBHOOK_TOKEN)

const log = (msg) => {
  console.log(msg)
  hook.send(new MessageEmbed().setDescription(msg).setTitle('MakePDF – Debug').setColor('#ED4539')).catch(() => {})
}
const lastRestart = Date.now()

client.on('message', async (msg) => {
  const { author, attachments, channel } = msg
  const { mentions, guild, member, content } = msg //required by debug command

  if (author.id == client.user.id) return

  if (content.includes('debug'))
    if (
      channel.type == 'dm' ||
      (mentions?.has(guild?.me, { ignoreEveryone: true }) && member?.hasPermission('ADMINISTRATOR'))
    )
      return msg.reply(
        new MessageEmbed().setTitle('MakePDF – Debug').setColor('#ED4539').setDescription(`
          **version :** MakePDF v${info.version}
          **time :** ${Date.now()}
          **lastRestart :** ${lastRestart}
          **guildId :** ${msg.guild?.id}
          **memberId :** ${author.id}
          **channelId :** ${channel.id}
        `)
      )

  const filesArray = attachments.array()

  filesArray.length > 0 &&
    filesArray.forEach(({ name, url }) => {
      if (!url) return

      const formats = process.env.MAKEPDF_SETTINGS_FORMATS.split(',')
      const extension = name.substring(name.lastIndexOf('.') + 1)

      if (!formats.includes(extension)) return

      get(url, (res) => {
        const bufs = []
        res.on('data', (chunk) => bufs.push(chunk))
        res.on('error', (err) => {
          channel.send(`Sorry, the conversion has failed :cry:`)
          return log(`Error during HTTP request : ${err.message}`)
        })
        res.on('end', () => {
          const fileData = Buffer.concat(bufs)

          convert(fileData, '.pdf', undefined, (err, pdfData) => {
            if (err) {
              channel.send(`Sorry, the conversion has failed :cry:`)
              return log(`Error converting file : ${err}`)
            }

            const newName = name.substring(0, name.lastIndexOf('.')) + '.pdf'
            const newAttachment = new MessageAttachment(pdfData, newName)
            channel.send(`**:paperclip: Here is your converted PDF file :**`, newAttachment)
          })
        })
      })
    })
})

const updateGuildCount = (server_count) => {
  client.user.setActivity(`${server_count} servers ⚡`, { type: 'WATCHING' })
  request({
    hostname: 'top.gg',
    port: 443,
    path: `/api/bots/${process.env.MAKEPDF_BOT_ID}/stats`,
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `${process.env.MAKEPDF_TOPGG_TOKEN}`
    }
  }).end(
    JSON.stringify({
      server_count
    })
  )
}

client.on('ready', () => {
  log(`Bot (re)started, version ${info.version}`)
  updateGuildCount(client.guilds.cache.size)
})
client.on('guildCreate', () => updateGuildCount(client.guilds.cache.size))
client.on('guildDelete', () => updateGuildCount(client.guilds.cache.size))

client.on('shardError', (e) => log(`Websocket connection error: ${e}`))
process.on(
  'unhandledRejection', //@ts-ignore
  (e) => e.code != 50013 && e.code != 50001 && log(`Unhandled promise rejection:\n\n${e.stack}\n\n${JSON.stringify(e)}`)
)

client.login(process.env.MAKEPDF_DISCORD_TOKEN)


VoiceNotify/voicenotify.js
/* eslint-disable no-case-declarations */
/* eslint-disable no-fallthrough */
/* eslint-disable eqeqeq */

import { Client, WebhookClient, MessageMentions, MessageEmbed } from 'discord.js'
import { initializeApp, cert } from 'firebase-admin/app'
import { getDatabase } from 'firebase-admin/database'
import { request } from 'https'

import info from './package.json'

const client = new Client()
const hook = new WebhookClient(process.env.VOICENOTIFY_WEBHOOK_ID, process.env.VOICENOTIFY_WEBHOOK_TOKEN)

const log = (msg) => {
  console.log(msg)
  hook.send(new MessageEmbed().setDescription(msg).setTitle('VoiceNotify – Debug').setColor('#08C754')).catch(() => {})
}
initializeApp({
  credential: cert({
    clientEmail: process.env.VOICENOTIFY_FIREBASE_CLIENT_EMAIL,
    privateKey: JSON.parse(`"${process.env.VOICENOTIFY_FIREBASE_PRIVATE_KEY}"`),
    projectId: process.env.VOICENOTIFY_FIREBASE_PROJECT_ID
  }),
  databaseURL: process.env.VOICENOTIFY_FIREBASE_DATABASE_URL
})
const db = getDatabase()
const lastRestart = Date.now()

const thresholdTimes = new Map() //last threshold time per channel
const broadcastTimes = new Map() //last broadcast time per channel

// settings management
const manager = {
  cache: {},
  get: async (g, c) => (c ? (await manager.get(g))?.[c] : (manager.cache[g] ||= (await db.ref(g).once('value')).val())),
  set: async (g, c, s) =>
    s &&
    (await db
      .ref(g)
      .child(c)
      .set(s)
      .then(async () => (await manager.get(g)) && (manager.cache[g][c] = s))),
  del: async (g, c) =>
    c
      ? await db
          .ref(g)
          .child(c)
          .remove()
          .then(() => delete manager.cache[g]?.[c])
      : await db
          .ref(g)
          .remove()
          .then(() => delete manager.cache[g])
}

client.on('voiceStateUpdate', async ({ channel: oldChannel }, { channel, guild }) => {
  // exit if user is leaving a channel
  if (!channel) return

  // exit if user changing mute/listen status
  if (channel.id == oldChannel?.id) return

  // fetch channel settings from db
  const settings = await manager.get(guild.id, channel.id)
  if (!settings) return

  // get text channel or delete if undefined (deleted channel)
  const textCh = await guild.channels.cache.find((ch) => ch.id == settings.text)
  if (!textCh?.isText() || textCh?.deleted) return manager.del(guild.id, channel.id)

  // exit if threshold is not reached
  if (channel.members.array().length < settings.min) return

  // get and set last threshold
  const lastThreshold = thresholdTimes.get(channel.id)
  thresholdTimes.set(channel.id, Date.now())

  // exit if threshold already reached recently
  // (progressive antispam depending on set threshold: 1p = 5m, 2p = 2.5m, 5p = 1m, 10p = 30s...)
  if (lastThreshold && Date.now() - lastThreshold < (5 / settings.min) * 60 * 1000) return

  // get last broadcast and exit if already sent <10m ago
  const lastBroadcast = broadcastTimes.get(channel.id)
  if (lastBroadcast && Date.now() - lastBroadcast < 10 * 60 * 1000) return

  // set last broadcast
  broadcastTimes.set(channel.id, Date.now())

  // get potential roles to mention
  const rolesList = settings.roles || ''

  // send message
  textCh.send(`\`🎙️\` A voice chat is taking place in "**${channel.name}**"!\n${rolesList}`)
})

client.on('message', async (msg) => {
  const { member, guild, mentions, content, channel } = msg

  if (!member || member.id == guild.me.id) return
  if (!mentions?.has(guild.me, { ignoreEveryone: true })) return
  if (!member.hasPermission('ADMINISTRATOR')) return msg.reply('you must be an administrator to use this bot.')

  const [, command, ...params] = content.toLowerCase().split(/ +/g)

  switch (command) {
    case 'enable' || 'disable':
      if (!member.voice.channelID) return msg.reply('you must be in a voice channel to use this bot.')

    case 'enable':
      const settings = {
        text: Number(channel.id),
        min: Number(/^\d+$/.test(params[0]) ? params[0] : '5'),
        roles: params?.toString().match(MessageMentions.ROLES_PATTERN)
      }
      await manager.set(guild.id, member.voice.channelID, settings)
      return msg.reply(
        `when ${settings.min} people or more are connected to "${
          member.voice.channel.name
        }", we will send an alert in <#${channel.id}> mentioning ${settings.roles?.length || '0'} role(s).`
      )

    case 'disable':
      await manager.del(guild.id, member.voice.channelID)
      return msg.reply(`notifications have been disabled for "${member.voice.channel.name}".`)

    case 'debug':
      msg.reply(
        new MessageEmbed().setTitle('VoiceNotify – Debug').setColor('#08C754').setDescription(`
              **version :** VoiceNotify v${info.version}
              **time :** ${Date.now()}
              **lastRestart :** ${lastRestart}
              **guildId :** ${guild.id}
              **memberId :** ${member.id}
              **textChannelId :** ${channel.id}
              **voiceChannelId :** ${member.voice?.channelID}
              **lastThreshold :** ${thresholdTimes.get(member.voice?.channelID)}
              **lastBroadcast :** ${broadcastTimes.get(member.voice?.channelID)}
              **guildSettings :**\n${JSON.stringify(await manager.get(guild.id))}
            `)
      )
      thresholdTimes.set(member.voice?.channelID, undefined)
      broadcastTimes.set(member.voice?.channelID, undefined)
      return

    default:
      return msg.reply(`
here are the bot commande to enable & disable voice chat notifications (administrators only) :

\`@VoiceNotify enable [threshold] [roles]\`
Enables voice chat notifications for the voice channel you are in, alerts will be sent to the channel where this command is executed.
Optional : [threshold] to trigger an alert defaults to 5 people ; [roles] will be mentioned when the alert is sent.

\`@VoiceNotify disable\`
Disables voice chat notifications for the voice channel you are in.
      `)
  }
})

const updateGuildCount = (server_count) => {
  client.user.setActivity(`${server_count} servers ⚡`, { type: 'WATCHING' })
  request({
    hostname: 'top.gg',
    port: 443,
    path: `/api/bots/${process.env.VOICENOTIFY_BOT_ID}/stats`,
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `${process.env.VOICENOTIFY_TOPGG_TOKEN}`
    }
  }).end(
    JSON.stringify({
      server_count
    })
  )
}

client.on('ready', () => {
  log(`Bot (re)started, version ${info.version}`)
  updateGuildCount(client.guilds.cache.size)
})
client.on('guildCreate', () => updateGuildCount(client.guilds.cache.size))
client.on('guildDelete', ({ id }) => {
  manager.del(id)
  updateGuildCount(client.guilds.cache.size)
})

client.on('shardError', (e) => log(`Websocket connection error: ${e}`))
process.on(
  'unhandledRejection', //@ts-ignore
  (e) => e.code != 50013 && e.code != 50001 && log(`Unhandled promise rejection:\n\n${e.stack}\n\n${JSON.stringify(e)}`)
)

client.login(process.env.VOICENOTIFY_DISCORD_TOKEN)
